File: FunctionsAsParameters/first_class_functions.py
"""
first_class_functions.py

Demonstrates the use of functions as parameters to other functions in Python.
(Also demonstrates lambdas and closures.)

GVSU 2025
"""

students = [
    {"first_name": "Ella", "last_name": "Smith", "gpa": 3.6, "credit_hours": 103},
    {"first_name": "Liam", "last_name": "Johnson", "gpa": 2.9, "credit_hours": 60},
    {"first_name": "Noah", "last_name": "Williams", "gpa": 3.8, "credit_hours": 15},
    {"first_name": "Olivia", "last_name": "Brown", "gpa": 3.2, "credit_hours": 75},
    {"first_name": "Emma", "last_name": "Jones", "gpa": 3.4, "credit_hours": 93},
    {"first_name": "Ava", "last_name": "Garcia", "gpa": 1.7, "credit_hours": 50},
    {"first_name": "Sophia", "last_name": "Martinez", "gpa": 2.8, "credit_hours": 80},
    {"first_name": "Isabella", "last_name": "Davis", "gpa": 3.5, "credit_hours": 60},
    {"first_name": "Mia", "last_name": "Rodriguez", "gpa": 1.0, "credit_hours": 40},
    {"first_name": "Amelia", "last_name": "Hernandez", "gpa": 3.9, "credit_hours": 105}
]

#
# Notice that the two functions below are nearly identical. 
# The only difference is the "if" condition.
#
def filter_for_probation(students, threshold):
    answer = []
    for s in students: 
        if s['gpa'] < threshold:
            answer.append(s)
    return answer

def filter_by_credit_hours(students, low, high):
    answer = []
    for s in students:
        if s['credit_hours'] >= low and s['credit_hours'] <= high:
            answer.append(s)
    return answer

#
# If we can pass code as a parameter, we can abstract the 
# two functions above:
#
def filter(students, filter_condition):
    answer = []
    for s in students: 
        if filter_condition(s):
            answer.append(s)
    return answer

def on_probation(s):
    return s['gpa'] < 2.0

def is_senior(s):
    return s['credit_hours'] >= 90

to_warn = filter(students, on_probation)
print(f"On probation: {to_warn}")

seniors = filter(students, is_senior)
print(f"Seniors: {seniors}")

#
# So far, so good.  But, notice that we are writing an entire function that 
# is only used once: as a parameter to another function. 
#
# Python allows you to create anonymous functions. These are called "lambdas"
#
to_warn = filter(students, lambda s : s['gpa'] < 2.0)
print(f"On probation: {to_warn}")

seniors = filter(students, lambda s : s['credit_hours'] >= 90)
print(f"Seniors: {seniors}")

juniors = filter(students, lambda s : s['credit_hours'] >= 60 and s['credit_hours'] < 90)
print(f"Juniors: {juniors}")

# In python, lambdas are limited to a single expression
# This is not true for most programming languages

#
# Closure 
#
# You can use the idea of a closure to build more complex filters.
# The "make_gpa_filter" function "closes" around the threshold 
# variable allowing you to effectively use it inside filter


def make_credit_hour_filter(low, high):
    return lambda s : s['credit_hours'] >= low and s['credit_hours'] < high

freshmen = filter(students, make_credit_hour_filter(0, 30))
sophomore = filter(students, make_credit_hour_filter(30, 60))
junior = filter(students, make_credit_hour_filter(60, 90))
senior = filter(students, make_credit_hour_filter(90, 10000))

print(f"Freshmen {freshmen}")
print(f"Sophomore {sophomore}")
print(f"Junior {junior}")
print(f"Senior {senior}")

File: FunctionsAsParameters/first_class_functions.js
const students = [
    { firstName: "Ella", lastName: "Smith", gpa: 3.6, creditHours: 103 },
    { firstName: "Liam", lastName: "Johnson", gpa: 2.9, creditHours: 60 },
    { firstName: "Noah", lastName: "Williams", gpa: 3.8, creditHours: 30 },
    { firstName: "Olivia", lastName: "Brown", gpa: 3.2, creditHours: 75 },
    { firstName: "Emma", lastName: "Jones", gpa: 3.4, creditHours: 93 },
    { firstName: "Ava", lastName: "Garcia", gpa: 1.7, creditHours: 50 },
    { firstName: "Sophia", lastName: "Martinez", gpa: 2.8, creditHours: 80 },
    { firstName: "Isabella", lastName: "Davis", gpa: 3.5, creditHours: 60 },
    { firstName: "Mia", lastName: "Rodriguez", gpa: 1.0, creditHours: 40 },
    { firstName: "Amelia", lastName: "Hernandez", gpa: 3.9, creditHours: 105 }
];


//
// Notice that the two functions below are nearly identical. 
// The only difference is the "if" condition.
//
function filterForProbation(students, threshold) {
    const answer = [];
    for (const student of students) {
        if (student.gpa < threshold) {
            answer.push(student);
        }
    }
    return answer;
}

function filterByCreditHours(students, low, high) {
    const answer = [];
    for (const student of students) {
        if (student.creditHours >= low && student.creditHours <= high) {
            answer.push(student);
        }
    }
    return answer;
}


//
// If we can pass code as a parameter, we can abstract the 
// two functions above:
//
function filter(theList, theFilter) {
    const answer = [];
    for (const item of theList) {
        if (theFilter(item)) {
            answer.push(item);
        }
    }
    return answer;
}

function isSenior(student) {
    return student.creditHours >= 90;
}

function onProbation(student) {
    return student.gpa < 2.0;
}

const seniors = filter(students, isSenior);
console.log(`Seniors:`, seniors);

const toWarn = filter(students, onProbation);
console.log(`On probation:`, toWarn);


//
// So far, so good.  But, notice that we are writing an entire function that 
// is only used once: as a parameter to another function. 
//
// Python allows you to create anonymous functions. These are called "lambdas"

seniors2 = filter(students, (s) => s.creditHours >= 90)
console.log(`Seniors 2:`, seniors2);

toWarn2 = filter(students, (s) => s.gpa < 2.0)
console.log(`On probation:`, toWarn2);

//
// JavaScript lambdas are *not* limited to a single expression.
//

gradedProbation = filter(students, (s) => {
    
    // You can't be on probation until 
    // you have completed 30 hours.
    if (s.creditHours < 30) {
        return false;
    }
    if (s.creditHours < 60) {
        return s.gpa < 2.0;
    }
    if (s.creditHours < 90) {
        return s.gpa < 2.25;
    }
    return s.gpa < 2.5
})
console.log(`On probation: ${gradedProbation}`);

//
// Closures
//
// You can use the idea of a closure to build more complex filters.

//

function makeGPAfilter(threshold) {
    return  (s) => s['gpa'] >= threshold;
}

const deansList = filter(students, (makeGPAfilter(3.0)))
console.log(`Students with 3.0 or better: `, deansList)

above_35 = filter(students, makeGPAfilter(3.5))
console.log(`Students with 3.5 or better:`, above_35)
